src/coreclr/vm/gchelpers.cpp

```diff
@@ -183,6 +183,76 @@ inline void CheckObjectSize(size_t alloc_size)
+#define g_SamplingDistributionMean (100 * 1024)
+
+// TODO: what are the call sites other than the fast path helpers?
+// if there is not,
+//      isSampled =
+//          (pEEAllocContext->alloc_sampling != nullptr) &&
+//          (pEEAllocContext->alloc_sampling != pAllocContext->alloc_limit);
+//
+inline Object* Alloc(ee_alloc_context* pEEAllocContext, size_t size, GC_ALLOC_FLAGS flags)
+{
+    CONTRACTL {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE; // returns an objref without pinning it => cooperative
+    } CONTRACTL_END;
+
+    Object* retVal = nullptr;
+    gc_alloc_context* pAllocContext = &pEEAllocContext->gc_alloc_context;
+
+    // TODO: isSampled could be an out parameter used by PublishObjectAndNotify()
+    //       where the event would be emitted
+    bool isSampled = false;
+    size_t samplingBudget = (size_t)(pAllocContext->alloc_limit - pEEAllocContext->alloc_sampling);
+    size_t availableSpace = (size_t)(pAllocContext->alloc_limit - pAllocContext->alloc_ptr);
+    if (flags & GC_ALLOC_USER_OLD_HEAP)
+    {
+        // TODO: if sampling is on, decide if this object should be sampled
+        // the provided allocation context fields are not used for LOH/POH allocations
+        // (only its alloc_bytes_uoh field will be updated)
+        // so get a random size in the distribution and if it is less than the size of the object
+        // then this object should be sampled
+    }
+    else
+    {
+        // check if this allocation overlaps the SOH sampling point:
+        //    if the sampling threashold was outside of the allocation context,
+        //    then
+        //      alloc_sampling was set to alloc_limit
+        //      or the allocation context was just initialized to nullptr for the first allocation
+        //    else
+        //      use the alloc_sampling to decide if this allocation should be sampled
+        // see the comment at the beginning of the function for the simplified check
+        isSampled =
+            (pEEAllocContext->alloc_sampling != nullptr) &&  // nullptr first time the allocation context is used
+            (pEEAllocContext->alloc_sampling != pAllocContext->alloc_limit) &&
+            (size > samplingBudget);
+    }
+
+    GCStress<gc_on_alloc>::MaybeTrigger(pAllocContext);
+
+    // for SOH, if there is enough space in the current allocation context, then
+    // the allocation will be done in place (like in the fast path),
+    // otherwise a new allocation context will be provided
+    retVal = GCHeapUtilities::GetGCHeap()->Alloc(pAllocContext, size, flags);
+
+    // only SOH allocations require sampling threshold to be recomputed
+    if ((flags & GC_ALLOC_USER_OLD_HEAP) == 0)
+    {
+        // the only case for which we don't need to recompute the sampling limit is when
+        // it was a fast path allocation (i.e. enough space left in the AC) after a sampled allocation
+        bool wasInPlaceAllocation = isSampled && (size <= availableSpace);
```

Comment:
There are a variety of conditions that need recomputing the sampling threshold:
1. if the object was sampled and it fit in the current AC
2. if the object was sampled and it didn't fit in the current AC
3. if the GC decided to move the AC for whatever reason
4. if the GC decided to resize the AC so that new_ac_size != availableSpace - size

Example (3) or (4) could happen in practice because some allocation flags cause the GC to create extra alignment padding. We should also try to avoid having the correctness of this logic depend on predicting how the GC will update an AC during the allocation.

I think the previous logic I had here was handling 1, 2, and 3 but missed 4.

Reply:
I'm not following your logic  :^)
Based on the earlier discussions, I thought that my idea of _accumulating allocations until the sampling threshold is reached and recompute a new one_ has been replaced by _computing such a sampling threshold each time a new allocation context is created_.

So, in this function, we only need to detect your scenario (2): the case of an allocation that triggers the slow path and crosses the limit of the allocation context (i.e. the GC provides a new allocation context). 

The scenario (1) should not recompute a new sampling limit because we are still in the same allocation context: we don't want to roll the dice another time for the same allocation context.
However, in that case, I should set alloc_sampling = alloc_limit

Reply:
When we use ComputeSamplingLimit() to produce a random number with geometric distribution, that tells us how many coin flips did it take to reach the first heads. Say ComputeSamplingLimit() returns 1000, that means we flipped 999 tails and then on the 1000'th time a heads. But the AC is 8000 bytes and we've only flipped 1000 coins so far, we can't just ignore the other 7000 bytes. We need to keep calling ComputeSamplingLimit() to simulate flipping more coins.

As an alternative thought experiment, setting alloc_sampling = alloc_limit after hitting the first sample means we will never have more than one sample in an AC. But if you flipped one coin for each byte in the AC the probability of getting two or more heads isn't zero. Any algorithm that samples at most once per AC can't be correct.

Reply:
Ok... so now, we compute a new threshold for allocations in SOH